### DISCUSSION TITLE
v5 `shallow` cares about set order.

### DISCUSSION THREAD

**Discussion by lakardion** [pmndrs/zustand#2808] [Bug report]
### Bug Description

Hello there. I was just playing around with this and noticed that the new shallow returns `false` for two sets that have the same elements inside.

```typescript
import { shallow } from "zustand/shallow";

const firstArray = [1, 2, 3];
const myFirstSet = new Set(firstArray);
const secondArray = [2, 1, 3];
const mySecondSet = new Set(secondArray);

console.log({
  myFirstSet,
  mySecondSet,
});

// v5 shallow returns false
const result = shallow(myFirstSet, mySecondSet);
```

Created two sandboxes as well to showcase that the previous version was returning `true`
v4 shallow:
https://codesandbox.io/p/sandbox/m8lkgp
v5 shallow
https://codesandbox.io/p/sandbox/shallow-check-zustand-zkjy77

### Reproduction Link

https://codesandbox.io/p/sandbox/shallow-check-zustand-zkjy77

**Comment by dbritto-dev** [comment]
@lakardion  thanks, I'll take care of that

**Reply by dai-shi** [reply]
@lakardion nice catch.

@dbritto-dev let me know if you need a help.

**Reply by dbritto-dev** [reply]
@dai-shi I came up with this idea -> https://stackblitz.com/edit/vitejs-vite-5fmphy?file=src%2Fshallow.ts

**Reply by dai-shi** [reply]
We should only ignore the order with `Set`.
`shallow([1, 2], [2, 1])` is false.

**Reply by dbritto-dev** [reply]
@dai-shi https://stackblitz.com/edit/vitejs-vite-5fmphy?file=src%2Fmain.ts,src%2Fshallow.ts

**Reply by dbritto-dev** [reply]
shall we ignore the order with `Map` too?

**Reply by dai-shi** [reply]
Map is already handled. I guess we should check `instanceof Set`.

**Reply by dbritto-dev** [reply]
I added an example for `Map` too, currently return `false`

**Reply by dbritto-dev** [reply]
<img width="1233" alt="image" src="https://github.com/user-attachments/assets/ac05fa97-242e-4be8-9f4f-3c1bedc4bf5c">

**Reply by dbritto-dev** [reply]
@lakardion  @dai-shi lmk your thoughts

**Reply by lakardion** [reply]
This looks good to me @dbritto-dev 🚀 !

**Reply by dbritto-dev** [reply]
@dai-shi any thoughts?

**Reply by dai-shi** [reply]
It seems very different from the current implementation. Does it pass all the tests? In any case, my idea to implement this is slightly different.

**Reply by dbritto-dev** [reply]
@dai-shi let me run some tests

**Reply by dbritto-dev** [reply]
@dai-shi https://github.com/pmndrs/zustand/pull/2816
